serialization: implement new serialization traits for the currently supported types

piodul opened the pull request:
The new `SerializeCql` and `SerializeRow` traits are supposed to eventually replace `Value` and `ValueList`. Currently, the new traits are implemented with a blanket implementation based on the old traits - this was just done in order to parallelize work on adapting the whole codebase to the new traits; the blanket implementations are inefficient and not type safe. This PR gets rid of them and, for each type for which we implement `Value` or `ValueList`, introduces proper implementations of `SerializeCql` and `SerializeRow`.

All types now check whether the CQL type serialized to matches, and most of them serialize exactly the same as they used to with the old traits. However, some slight differences were made in order to utilize the information available to the new traits and make it more robust:

- `CqlValue::UserDefinedType` now doesn't assume that the order of the fields is the same as in the CQL type - it sorts the fields according to the CQL type.
- `impl SerializeRow for BTreeMap/HashMap` doesn't serialize in the "named values" format anymore as it is no longer supported by the new traits. Instead, it uses provided information about the column names to properly order the values before sending them to the database.

Existing tests in `value_tests.rs` are adapted so that they perform serialization using both old and new traits and compare the results (with a small number of exceptions where serialization behavior was changed).

Fixes: https://github.com/scylladb/scylla-rust-driver/issues/821

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

piodul commented:
v1.1: more complete docstrings for the error types

Lorak-mmk commented:
> 
> Depends on #851

Isn't this the other way around?

piodul commented:
> > Depends on #851
> 
> Isn't this the other way around?

Yeah, my mistake. Updated the description.

piodul commented:
v2:
- rebased
- addressed most of the comments that were placed in https://github.com/scylladb/scylla-rust-driver/pull/851 (apart from the one about panicking on overflow)

Lorak-mmk commented on scylla-cql/src/types/serialize/value.rs:
```diff
@@ -29,19 +48,883 @@ pub trait SerializeCql {
+
+    let rust_name = std::any::type_name::<CqlValue>();
+
+    match Arc::get_mut(&mut err) {
+        Some(err_mut) => {
+            if let Some(err) = err_mut.downcast_mut::<BuiltinTypeCheckError>() {
+                err.rust_name = rust_name;
+            } else if let Some(err) = err_mut.downcast_mut::<BuiltinSerializationError>() {
+                err.rust_name = rust_name;
+            }
+        }
+        None => {
+            // The `None` case shouldn't happen consisdering how we are using
+            // the function in the code now, but let's provide it here anyway
+            // for correctness.
+            if let Some(err) = err.downcast_ref::<BuiltinTypeCheckError>() {
+                if err.rust_name != rust_name {
+                    return Arc::new(BuiltinTypeCheckError {
+                        rust_name,
+                        ..err.clone()
+                    });
+                }
+            }
+            if let Some(err) = err.downcast_ref::<BuiltinSerializationError>() {
+                if err.rust_name != rust_name {
+                    return Arc::new(BuiltinSerializationError {
+                        rust_name,
+                        ..err.clone()
+                    });
+                }
+            }
+        }
+    };
```
Oh wait, I didn't read the code properly, sorry.

piodul commented:
v3:
- adjusted the builder interface to make it safer/easier to handle CQL value size overflows
- adjusted the `impl SerializeCql` blocks to the new interface - new errors are handled by unwraps for fixed sized types, and by properly returning an error for dynamically sized ones
- got rid of the `.expect` calls in map/list/set serialization code

piodul commented:
Looks like the CI failed due to a known issue: https://github.com/scylladb/scylla-rust-driver/issues/813

Lorak-mmk commented on scylla-cql/src/types/serialize/row.rs:
```diff
@@ -50,22 +55,310 @@ pub trait SerializeRow {
     }
 }
 
+impl SerializeRow for SerializedValues {
+    fallback_impl_contents!();
+}
+
+impl<'b> SerializeRow for Cow<'b, SerializedValues> {
+    fallback_impl_contents!();
+}
```
Is this a implementation created to make transition to new API easier? If so, let's add a TODO / create and issue so that we don't forget to remove it later.

Lorak-mmk commented on scylla-cql/src/types/serialize/value.rs:
```diff
@@ -29,19 +48,940 @@ pub trait SerializeCql {
+        .map_err(|_| mk_ser_err_named(rust_name, typ, BuiltinSerializationErrorKind::SizeOverflow))
+}
 
 pub fn serialize_legacy_value<T: Value, W: CellWriter>(
```
Is this used anywhere after the changes in this PR? I can't find any references to it. If not, it can be removed

piodul commented:
v4:
- Rebased
- Implemented two declarative macros that allows to generate a quick and dirty implementation of SerializeRow/SerializeCql based on ValueList/Value
